### PR TITLE
access log: add DYNAMIC_METADATA_UNQUOTED formatter

### PR DESCRIPTION
**Pull request by rgs1**
This is useful when you are using the header-to-metadata filter to clean
up headers and then you want to send that to the access log as plain, non
quoted, text.

Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

### CODE REVIEW

**Comment by rgs1** [comment]
Per @fishcakez , `DYNAMIC_METADATA_TRIM()` is possibly a better name for this formatter. Thoughts?

**Comment by rgs1** [comment]
> Per @fishcakez , `DYNAMIC_METADATA_TRIM()` is possibly a better name for this formatter. Thoughts?

FWIW, I prefer `DYNAMIC_METADATA_UNQUOTED()` since we are already using it like that :-)

File: source/common/formatter/substitution_formatter.cc [hunk 1]
```diff
@@ -980,8 +990,9 @@ GrpcStatusFormatter::formatValue(const Http::RequestHeaderMap&,
 
 MetadataFormatter::MetadataFormatter(const std::string& filter_namespace,
                                      const std::vector<std::string>& path,
-                                     absl::optional<size_t> max_length)
-    : filter_namespace_(filter_namespace), path_(path), max_length_(max_length) {}
+                                     absl::optional<size_t> max_length, const bool unquoted)
```
Comment (zuercher) [inline source/common/formatter/substitution_formatter.cc:993]: nit: I don't think we usually use const parameters that aren't references or pointers.

File: source/common/formatter/substitution_formatter.h [hunk 2]
```diff
@@ -318,6 +318,7 @@ class MetadataFormatter {
   std::string filter_namespace_;
   std::vector<std::string> path_;
   absl::optional<size_t> max_length_;
+  const bool unquoted_{};
```
Comment (zuercher) [inline source/common/formatter/substitution_formatter.h:321]: Since it's always initialized in the constructor, I think you can remove `{}`.

**Review by zuercher — COMMENTED** [review]
I have to say I'm leery of this change. Feels like we're in this perpetual state of adding more and more little tweeks for this format or that.

**Comment by rgs1** [comment]
> I have to say I'm leery of this change. Feels like we're in this perpetual state of adding more and more little tweeks for this format or that.

I know. I am also not super excited about having to add this, but I am not sure what other options we have.

There's no way to add internal formatters and the cleanest way for us to scrub/clean headers that needed to be logged is:

* header-to-metadata -> apply regex
* DYNAMIC_METADATA

but the latter isn't happy with old school logging (e.g.: plain)....

File: source/common/formatter/substitution_formatter.cc (hunk 1 above)
Comment (rgs1) [reply source/common/formatter/substitution_formatter.cc:993]: not in headers, but it's used on definitions here and there:

```
 $ git grep 'const bool ' -- '*.cc' | grep :: | grep -v test/ | grep -v ';'
source/common/formatter/substitution_formatter.cc:                                     absl::optional<size_t> max_length, const bool unquoted)
source/extensions/filters/common/ext_authz/check_request_utils.cc:                                           const std::string& service, const bool local,
source/extensions/filters/network/redis_proxy/command_splitter_impl.cc:void SplitRequestBase::updateStats(const bool success) {
source/extensions/filters/network/thrift_proxy/router/router_impl.cc:void Router::UpstreamRequest::releaseConnection(const bool close) {
source/extensions/filters/network/zookeeper_proxy/filter.cc:void ZooKeeperFilter::onConnect(const bool readonly) {
source/extensions/filters/network/zookeeper_proxy/filter.cc:void ZooKeeperFilter::onGetDataRequest(const std::string& path, const bool watch) {
source/extensions/filters/network/zookeeper_proxy/filter.cc:void ZooKeeperFilter::onGetChildrenRequest(const std::string& path, const bool watch,
source/extensions/filters/network/zookeeper_proxy/filter.cc:void ZooKeeperFilter::onExistsRequest(const std::string& path, const bool watch) {
source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc:    const bool ok = Network::Socket::applyOptions(
source/extensions/quic_listeners/quiche/active_quic_listener.cc:    const bool ok = Network::Socket::applyOptions(
source/extensions/stat_sinks/common/statsd/statsd.cc:                             Network::Address::InstanceConstSharedPtr address, const bool use_tag,
source/extensions/tracers/zipkin/span_buffer.cc:JsonV2Serializer::JsonV2Serializer(const bool shared_span_context)
source/extensions/tracers/zipkin/span_buffer.cc:ProtobufSerializer::ProtobufSerializer(const bool shared_span_context)
source/server/admin/prometheus_stats.cc:    Buffer::Instance& response, const bool used_only, const absl::optional<std::regex>& regex,
source/server/admin/prometheus_stats.cc:    const bool used_only, const absl::optional<std::regex>& regex) {
source/server/admin/stats_handler.cc:                          const bool used_only, const absl::optional<std::regex> regex,
```

but yeah not super popular.

**Comment by zuercher** [comment]
I think it should be possible to allow extensions to provide their own formatters. But maybe that's one extension type too many.

**Comment by rgs1** [comment]
@zuercher so shall we let this one land or do we want to explore the extensions path?

**Comment by mattklein123** [comment]
FWIW we already have an extension point for access log filters. Having one for formatters seems like it might be worth it also.

**Comment by zuercher** [comment]
The only problem I have with landing it that it's forever. I think I'd prefer some change to the parsing of `DYNAMIC_METADATA(NAMESPACE:KEY)` that triggered quote stripping.

**Comment by rgs1** [comment]
> The only problem I have with landing it that it's forever. I think I'd prefer some change to the parsing of `DYNAMIC_METADATA(NAMESPACE:KEY)` that triggered quote stripping.

Alas, I don't think it would be pretty because:

* `namespace` _should_ have reverse DNS format [0], but we don't really enforce that so we can't assume anything
* `key` is really free form, so same thing

which means we can't -- cleanly -- add more parameters within the parentheses.

After the final `)`, we must have the max length param. We could make it such that 0 means unset, so
if you wanted an extra param -- without setting max length -- you'd end with:

```
DYNAMIC_METADATA(envoy.lb:foo.bar):0,true
```

to get what we want.... which is a bit yucky. 

Hmm. Thoughts?

[0] https://github.com/envoyproxy/envoy/blob/master/include/envoy/network/filter.h#L276

**Comment by rgs1** [comment]
> FWIW we already have an extension point for access log filters. Having one for formatters seems like it might be worth it also.

Sounds like this would be the way forward then, given my above comment about extending `DYNAMIC_METADATA()`...

**Comment by zuercher** [comment]
I think moving to extensions would be the best thing. 

For the suffix to control output, you could have a further :-separated list of things. So `:10:strip` would limit length to 10 and strip quotes or just `:strip` would strip quotes without affecting the length.

**Comment by rgs1** [comment]
Closing this in favor of #13411.

**Pull request closed without merging**